src/wui/info_panel.cc

```diff
@@ -0,0 +1,518 @@
+	last_mouse_pos_ = Vector2i(x, y);
+	return is_mouse_over_panel();
+}
+
+size_t InfoPanel::index_of(const MessagePreview& mp) const {
+	size_t i = 0;
+	for (const MessagePreview* m : messages_) {
+		if (m == &mp) {
+			return i;
+		}
+		++i;
+	}
+	NEVER_HERE();
+}
+
+void InfoPanel::log_message(const std::string& message) {
+	push_message(*new MessagePreview(*this, message, ""));
+}
+
+void InfoPanel::pop_message(MessagePreview& m) {
+	m.set_visible(false);
+	m.die();
+	for (auto it = messages_.begin(); it != messages_.end(); ++it) {
+		if (*it == &m) {
+			messages_.erase(it);
+			layout();
+			return;
+		}
+	}
+	NEVER_HERE();
+}
+
+void InfoPanel::push_message(MessagePreview& message) {
+	messages_.push_back(&message);
+
+	if (messages_.size() > kMaxMessages) {
+		pop_message(*messages_.front());
+	}
+
+	layout();
+}
+
+void InfoPanel::set_fps_string(const std::string& t) {
+	text_fps_.set_text(t);
+}
+void InfoPanel::set_coords_string(const std::string& t) {
+	text_coords_.set_text(t);
+}
+void InfoPanel::set_time_string(const std::string& t) {
+	time_string_ = t;
+	update_time_speed_string();
+}
+void InfoPanel::set_speed_string(const std::string& t) {
+	speed_string_ = t;
+	update_time_speed_string();
+}
+
+void InfoPanel::update_time_speed_string() {
+	std::string realtime;
+	if (draw_real_time_) {
+		std::time_t t = std::time(nullptr);
+		std::tm tm = *std::localtime(&t);
+		std::ostringstream oss("");
+		oss << std::put_time(&tm, "%X");
+		realtime = oss.str();
+	}
+
+	std::vector<std::string*> non_empty;
+	std::string format;
+	for (std::string* s : {&time_string_, &realtime, &speed_string_}) {
+		if (!s->empty()) {
+			non_empty.push_back(s);
+		}
+	}
+
+	boost::format f;
+	switch (non_empty.size()) {
+	case 0:
+		text_time_speed_.set_text("");
+		return;
+	case 1:
+		text_time_speed_.set_text(*non_empty.back());
+		return;
+	case 2:
+		f = boost::format(_("%1$s · %2$s"));
+		break;
+	case 3:
+		f = boost::format(_("%1$s · %2$s · %3$s"));
+		break;
+	default:
+		NEVER_HERE();
+	}
+
+	for (std::string* s : non_empty) {
+		f % *s;
+	}
+	text_time_speed_.set_text(f.str());
+}
+
+void InfoPanel::fast_forward_message_queue() {
+	iplayer_ = dynamic_cast<InteractivePlayer*>(&ibase_);
+	assert(iplayer_);
+	last_message_id_.reset(
+	   new Widelands::MessageId(iplayer_->player().messages().current_message_id()));
+}
+
+void InfoPanel::think() {
+	if (!last_message_id_) {
+		last_message_id_.reset(new Widelands::MessageId());
+		iplayer_ = dynamic_cast<InteractivePlayer*>(&ibase_);
+	}
+	if (iplayer_ && !message_queue_) {
+		message_queue_ = &iplayer_->player().messages();
+	}
+
+	while (message_queue_ && *last_message_id_ != message_queue_->current_message_id()) {
+		*last_message_id_ = Widelands::MessageId(last_message_id_->value() + 1);
+		if (message_queue_->count(last_message_id_->value())) {
+			push_message(
+			   *new MessagePreview(*this, *(*message_queue_)[*last_message_id_], *last_message_id_));
```

Comment:
It's like Christmas: stars everywhere ;). Maybe use pointer/literal variables instead of references in this case.
`last_message_id_ -> Widelands::MessageId `
`push_message -> push_message(MessagePreview*)`
`pop_message -> pop_message(MessagePreview*)`
This will also allow you to remove some other * and &.